contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,591 @@
+    while (nextJob - ctx->jobWriteID >= ctx->numJobs) {
+        ctx->stats.waitWrite++;
+        ctx->stats.writeCounter++;
+        DEBUGLOG(2, "waiting on job Write, nextJob: %u\n", nextJob);
+        pthread_cond_wait(&ctx->jobWrite_cond, &ctx->jobWrite_mutex);
+    }
+    pthread_mutex_unlock(&ctx->jobWrite_mutex);
+    DEBUGLOG(2, "createCompressionJob(): continuing after job write\n");
+
+
+    job->compressionLevel = ctx->compressionLevel;
+    job->src.start = malloc(srcSize);
+    job->src.size = srcSize;
+    job->dst.size = ZSTD_compressBound(srcSize);
+    job->dst.start = malloc(job->dst.size);
+    job->jobID = nextJob;
+    if (!job->src.start || !job->dst.start) {
+        /* problem occurred, free things then return */
+        DISPLAY("Error: problem occurred during job creation\n");
+        free(job->src.start);
+        free(job->dst.start);
+        return 1;
+    }
+    memcpy(job->src.start, data, srcSize);
+    pthread_mutex_lock(&ctx->jobReady_mutex);
+    ctx->jobReadyID++;
+    pthread_cond_signal(&ctx->jobReady_cond);
+    pthread_mutex_unlock(&ctx->jobReady_mutex);
+    DEBUGLOG(2, "finished job creation %u\n", nextJob);
+    ctx->nextJobID++;
+    return 0;
+}
+
+static void printStats(cStat_t stats)
+{
+    DISPLAY("========STATISTICS========\n");
+    DISPLAY("# times waited on job ready: %u\n", stats.waitReady);
+    DISPLAY("# times waited on job compressed: %u\n", stats.waitCompressed);
+    DISPLAY("# times waited on job Write: %u\n\n", stats.waitWrite);
+}
+
+static int compressFilename(const char* const srcFilename, const char* const dstFilenameOrNull)
+{
+    BYTE* const src = malloc(FILE_CHUNK_SIZE);
+    unsigned const stdinUsed = !strcmp(srcFilename, stdinmark);
+    FILE* const srcFile = stdinUsed ? stdin : fopen(srcFilename, "rb");
+    const char* const outFilenameIntermediate = (stdinUsed && !dstFilenameOrNull) ? stdoutmark : dstFilenameOrNull;
+    const char* outFilename = outFilenameIntermediate;
+    char fileAndSuffix[MAX_PATH];
+    size_t const numJobs = MAX_NUM_JOBS;
+    int ret = 0;
+    adaptCCtx* ctx = NULL;
+    UTIL_getTime(&g_startTime);
+    g_streamedSize = 0;
+
+    if (!outFilenameIntermediate) {
+        if (snprintf(fileAndSuffix, MAX_PATH, "%s.zst", srcFilename) + 1 > MAX_PATH) {
+            DISPLAY("Error: output filename is too long\n");
+            ret = 1;
+            goto cleanup;
+        }
+        outFilename = fileAndSuffix;
+    }
+
+    /* checking for errors */
+    if (!srcFilename || !outFilename || !src || !srcFile) {
+        DISPLAY("Error: initial variables could not be allocated\n");
+        ret = 1;
+        goto cleanup;
+    }
+
+    /* creating context */
+    ctx = createCCtx(numJobs, outFilename);
+    if (ctx == NULL) {
+        ret = 1;
+        goto cleanup;
+    }
+
+    /* create output thread */
+    {
+        pthread_t out;
+        if (pthread_create(&out, NULL, &outputThread, ctx)) {
```

Comment:
I think you will sleep forever in `waitUntilAllJobsCompleted()` if this fails (and same for the compression thread). I think that applies for a lot of the failure conditions.